blog/2024-11-26-indexing-and-reorgs.md

```diff
@@ -0,0 +1,131 @@
+
+
+## Chain Reorgs & Independent Data
+
+Independent data is data that **doesn’t require the previous state in order to process the current state**. In an indexer, this can quite simply be contextualized as only “create” operations. 
+
+Considering you are ingesting data and the only CRUD (Create, Read, Update & Delete) operations your handlers perform are writing entities, then your indexer is handling stateless data, and the concept of handling reorgs is very simple. Effectively delete all the entities from orphaned blocks and re-ingest them with the canonical blocks data. This is possible since stateless data allows for parallel processing.
+
+Practically speaking we need to perform some form of aggregation on this data to process it into information, in this context our indexer is dealing with stateful data.
+
+> Side tangent: Stateless indexing can be handled incredibly fast by parallelization, such as indexers like [Flair](flair.dev) who achieve impressive speed with only RPC.  
+
+
+## Chain Reorgs and Stateful Data
+
+Stateful data is data that **depends on the previous state**, think update & delete, in order to process the current state. In an indexer, this means that operations such as updates and deletes are required, rather than simply writing new entities. When handling stateful data, your indexer needs to account for the current state of entities, and thus, reorgs become notably more complex.
+
+During a chain reorg, rather than simply replacing orphaned data, you need to **revert previous operations or changes** and ensure that the entity state is rolled back correctly. This requires tracking the history of changes to each entity so that, when a reorg occurs, you can accurately undo or adjust state based on the adopted canonical chain’s data.
+
+> Note: We can periodically prune the entity history, retaining only the changes relevant to unfinalized blocks.
+
+
+## Reorgs and Multichain Indexing
+
+The concept of reorgs and multichain indexing should be considered since during multichain indexing we’re adding another layer of complexity in indexing in the sense that we are ingesting from multiple sources in parallel. If we are indexing in [unordered_multichain_mode](https://docs.envio.dev/docs/HyperIndex/multichain-indexing#unordered-multichain-mode) then we process reorgs in the same way as we do with stateful data; however, in the context of multichain ordering, we roll the entity state back for all chains at the occurrence of a reorg on any of the given chains and re-process the events across all the networks from the block timestamp of the forked block 
+
+> Side tangent: The logic used in reorgs whereby we are able to revert individual entities state and reingest at a point in time can be used to implement a concept we’re calling optimistic head multichain indexing allows for real-time ordered indexing without waiting for the block of the slowest network to be mined. 
```

Comment:
Hey @DenhamPreen let me know what you think of this. Short and sweet but can expand it further if need be.

```suggestion
When it comes to multichain indexing, we face additional complexities as we process events from multiple sources that interact and update the same entity state. When one chain undergoes a reorg, we need to roll back the state to a known correct point and reprocess any events from all chains that affected the state after the reorg on the affected chain.
```